Mathlib/Tactic/LinearCombination.lean

```diff
@@ -115,24 +115,26 @@ def elabLinearCombination (tk : Syntax)
         from the term"
       `(Eq.refl 0)
     | .proof p => pure p
-  let norm := norm?.getD (Unhygienic.run <| withRef tk `(tactic| ring1))
-  let lem : Ident ← mkIdent <$> do
-    try
-      -- if we are in a "true" ring, with well-behaved negation, it is better to present the
-      -- normalization tactic with a goal of the form `[stuff] = 0`, because this gives more useful
-      -- error messages on failure
-      let _ ← synthInstance (← mkAppM ``Neg #[ty])
-      pure ``eq_of_sub
-    catch _ =>
-      -- but otherwise (for example over `ℕ` or `ℝ≥0`) we can solve the problem by presenting the
-      -- normalization tactic with a goal of the form `[stuff] = [stuff]`
-      pure ``eq_of_add
-  Term.withoutErrToSorry <| Tactic.evalTactic <| ← withFreshMacroScope <|
-  match exp? with
-  | some n =>
-    if n.getNat = 1 then `(tactic| (refine $lem $p ?a; case' a => $norm:tactic))
-    else `(tactic| (refine eq_of_add_pow $n $p ?a; case' a => $norm:tactic))
-  | _ => `(tactic| (refine $lem $p ?a; case' a => $norm:tactic))
+  -- build the term for the central `refine` in `linear_combination`
+  let p' ← do
+    match exp? with
+    | some n =>
+      if n.getNat = 1 then
+        `(eq_of_add $p ?a)
+      else
+        `(eq_of_add_pow $n $p ?a)
+    | _ => `(eq_of_add $p ?a)
+  -- run the central `refine` in `linear_combination`
+  Term.withoutErrToSorry <| Tactic.refineCore p' `refine false
+  -- if we are in a "true" ring, with well-behaved negation, we rearrange from the form
+  -- `[stuff] = [stuff]` to the form `[stuff] = 0`,
+  -- because this gives more useful error messages on failure
+  let _ ← Tactic.tryTactic <| Tactic.liftMetaTactic fun g ↦ g.applyConst ``eq_rearrange
+  -- now run the normalization tactic provided, or the default normalization if none is provided
+  match norm? with
+  | some norm => Tactic.evalTactic norm
+  | none => withRef tk <| Tactic.liftMetaFinishingTactic <|
+    fun g ↦ AtomM.run .instances <| Ring.proveEq g
```

Comment:
Could you add a comment that this is `ring1` but with `.instances` instead of `.reducible`? (And why it's not the same as `ring1`?)

bors d+